mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoSafeUpdateProtocol.java

```diff
@@ -0,0 +1,349 @@
+                        } else {
+                            results.put(we.getIndex(),
+                                        Error.get("update", MongoCrudConstants.ERR_SAVE_ERROR, we.getMessage()));
+                        }
+                    }
+                }
+                findConcurrentModifications(results);
+            }
+        }
+        reset();
+        return results;
+    }
+
+    /**
+     * This executes a query to find out documents with concurrent modification errors
+     */
+    protected void findConcurrentModifications(Map<Integer,Error> results) {
+        List<Object> updatedIds=new ArrayList<>(idsInBatch.size());
+        // Collect all ids without errors
+        int index=0;
+        for(Object id:idsInBatch) {
+            if(!results.containsKey(index))
+                updatedIds.add(id);
+            index++;
+        }
+        LOGGER.debug("checking for concurrent modifications:{}",updatedIds);
+        if(!updatedIds.isEmpty()) {
+            // documents with the given _ids and whose docver contains our docVer are the ones we managed to update
+            // others are failures
+            BasicDBObject query=new BasicDBObject(DOCVER_FLD,docVer);
+            query.append("_id",new BasicDBObject("$in",updatedIds));
+            DBCursor cursor=null;
+            try {
+                Set<Object> successfulIds=new HashSet<>(updatedIds.size());
+                cursor=collection.find(query,new BasicDBObject("_id",1));
+                while(cursor.hasNext()) {
+                    DBObject obj=cursor.next();
+                    Object id=obj.get("_id");
+                    successfulIds.add(id);
+                }
+                index=0;
+                for(Object id:idsInBatch) {
+                    if(!results.containsKey(index)) { // No other errors for this id
+                        if(!successfulIds.contains(id)) {
+                            // concurrency errors for this id
+                            results.put(index,Error.get("update",MongoCrudConstants.ERR_CONCURRENT_UPDATE,id.toString()));
+                        }
+                    }
+                    index++;
+                }
+            } finally {
+                if(cursor!=null)
+                    cursor.close();
+            }
+        }
+    }
+
+    /**
+     * Writes a replace query, which is:
+     * <pre>
+     *  {_id:<docId>, "@mongoHidden.docver.0":ObjectId(<originalDocver>)}
+     * </pre>
+     */
+    private DBObject writeReplaceQuery(DBObject doc) {
+        DBObject hidden=getHidden(doc,false);
+        ObjectId top=null;
+        if(hidden!=null) {
+            List<ObjectId> list=(List<ObjectId>)hidden.get(DOCVER);
+            if(list!=null&&!list.isEmpty())
+                top=list.get(0);
+        }
+        BasicDBObject query=new BasicDBObject("_id",doc.get("_id"));
+        query.append(DOCVER_FLD0,top);
```

Comment:
it is always used by _id, within a set of 'batchsize' docs, so index should
not be necessary.

On Fri, Aug 26, 2016 at 1:07 PM, Alec Henninger <notifications@github.com>
wrote:

> In mongo/src/main/java/com/redhat/lightblue/mongo/crud/
> MongoSafeUpdateProtocol.java
> <https://github.com/lightblue-platform/lightblue-mongo/pull/303#discussion_r76471533>
> :
>
> > +    /**
> > +     * Writes a replace query, which is:
> > +     * <pre>
> > +     *  {_id:<docId>, "@mongoHidden.docver.0":ObjectId(<originalDocver>)}
> > +     * </pre>
> > +     */
> > +    private DBObject writeReplaceQuery(DBObject doc) {
> > +        DBObject hidden=getHidden(doc,false);
> > +        ObjectId top=null;
> > +        if(hidden!=null) {
> > +            List<ObjectId> list=(List<ObjectId>)hidden.get(DOCVER);
> > +            if(list!=null&&!list.isEmpty())
> > +                top=list.get(0);
> > +        }
> > +        BasicDBObject query=new BasicDBObject("_id",doc.get("_id"));
> > +        query.append(DOCVER_FLD0,top);
>
> Should docver array be indexed by default?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/lightblue-platform/lightblue-mongo/pull/303/files/6e592caaaf4c79fd821914fdb9640896fa145ab0#r76471533>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADgDDdO43e5lxq72-ffNqs1XqO6M1KNeks5qjzlZgaJpZM4JtbtZ>
> .
>